agent/consul/leader_connect_ca_test.go

```diff
@@ -0,0 +1,256 @@
+	switch method {
+	case "ConnectCA.Roots":
+		roots := reply.(*structs.IndexedCARoots)
+		roots.TrustDomain = connect.TestClusterID
+		roots.Roots = []*structs.CARoot{m.primaryRoot}
+		roots.ActiveRootID = m.primaryRoot.ID
+	case "ConnectCA.SignIntermediate":
+		r := reply.(*string)
+		*r = m.primaryRoot.RootCert
+	}
+
+	m.callbackCh <- fmt.Sprintf("forwardDC/%s", method)
+
+	return nil
+}
+
+func (m *mockCAServerDelegate) generateCASignRequest(csr string) *structs.CASignRequest {
+	return &structs.CASignRequest{
+		Datacenter: m.config.PrimaryDatacenter,
+		CSR:        csr,
+	}
+}
+
+func (m *mockCAServerDelegate) raftApply(t structs.MessageType, msg interface{}) (interface{}, error) {
+	if t == structs.ConnectCARequestType {
+		req := msg.(*structs.CARequest)
+		act, err := m.store.CARootSetCAS(1, req.Index, req.Roots)
+		require.NoError(m.t, err)
+		require.True(m.t, act)
+
+		act, err = m.store.CACheckAndSetConfig(1, req.Config.ModifyIndex, req.Config)
+		require.NoError(m.t, err)
+		require.True(m.t, act)
+	}
+	m.callbackCh <- fmt.Sprintf("raftApply/%s", t)
+	return nil, nil
+}
+
+// mockCAProvider mocks an empty provider implementation with a channel in order to coordinate
+// waiting for certain methods to be called.
+type mockCAProvider struct {
+	callbackCh chan string
+	rootPEM    string
+}
+
+func (m *mockCAProvider) Configure(cfg ca.ProviderConfig) error { return nil }
+func (m *mockCAProvider) State() (map[string]string, error)     { return nil, nil }
+func (m *mockCAProvider) GenerateRoot() error                   { return nil }
+func (m *mockCAProvider) ActiveRoot() (string, error)           { return m.rootPEM, nil }
+func (m *mockCAProvider) GenerateIntermediateCSR() (string, error) {
+	m.callbackCh <- "provider/GenerateIntermediateCSR"
+	return "", nil
+}
+func (m *mockCAProvider) SetIntermediate(intermediatePEM, rootPEM string) error {
+	m.callbackCh <- "provider/SetIntermediate"
+	return nil
+}
+func (m *mockCAProvider) ActiveIntermediate() (string, error)                       { return m.rootPEM, nil }
+func (m *mockCAProvider) GenerateIntermediate() (string, error)                     { return "", nil }
+func (m *mockCAProvider) Sign(*x509.CertificateRequest) (string, error)             { return "", nil }
+func (m *mockCAProvider) SignIntermediate(*x509.CertificateRequest) (string, error) { return "", nil }
+func (m *mockCAProvider) CrossSignCA(*x509.Certificate) (string, error)             { return "", nil }
+func (m *mockCAProvider) SupportsCrossSigning() (bool, error)                       { return false, nil }
+func (m *mockCAProvider) Cleanup() error                                            { return nil }
+
+func waitForCh(t *testing.T, ch chan string, expected string) {
+	select {
+	case op := <-ch:
+		if op != expected {
+			t.Fatalf("got unexpected op %q, wanted %q", op, expected)
+		}
+	case <-time.After(3 * time.Second):
```

Comment:
You could instead use the current time and the time portion of `(*testing.T).Deadline()` instead to compute remaining test time instead of using a made up "not forever" number here and several other places